### ISSUE TITLE
[FEATURE] Approximate Nearest Neighbours for Recommender Systems

### ISSUE THREAD

**Issue by miguelgfierro** [microsoft/recommenders#822]
### Description

Add KNN for topk

Here there is a benchmark of KNN libs: https://www.benfrederickson.com/approximate-nearest-neighbours-for-recommender-systems/

**Comment by loomlike** [comment]
@gramhagen I can add item, user knns If you're not working on this

**Comment by gramhagen** [comment]
yeah go for it, I haven't had much time to work on anything honestly. were you thinking of doing approximate or exact knn?

**Comment by miguelgfierro** [comment]
In the link I added there is a comparison between different libraries for knn and a paper by Noam is cited. Maybe it would be good to ask him for some advice on what library to use

**Comment by loomlike** [comment]
@miguelgfierro  Checked a couple of options and probably `faiss` will be the best fit (in terms or ease of install + license). Will check Noam as well.

**Comment by loomlike** [comment]
@gramhagen approximate. RAPIDS also has knn based on faiss working on multi-gpu. That could be another option.
Btw, faiss doesn't support sparse vector. For large-size data, we should consider other options...

**Comment by gramhagen** [comment]
rapids sounds good. Also, @mdekstrand mentioned the knn implementation in [Python LensKit](https://github.com/lenskit/lkpy)  is significantly optimized to allow it to scale to fairly large datasets, so that's probably a good place to start.
